File: crawler/crawler.py
# -*- coding: utf-8 -*-
"""
Created on Wed Feb 19 17:16:23 2020

@author: billstark
"""

from typing import Union
from sys_utils import pdump, pload, backup_file
from utils import *
import extractor
from driver import DriverWrapper

from bs4 import BeautifulSoup
from tqdm import tqdm

addr_default = 'http://www.ocw.titech.ac.jp/index.php'

try:
  assert driver
except:
  driver = DriverWrapper(
    driver_path='./chromedriver104.exe', 
    mem_reset_limit=300
  )

# args
this_year = 2022
innovation_year = 2016

args_lang = lambda lang='JA': {'lang': lang}
args_year = lambda year=this_year: {'Nendo': year}
args_term = lambda zenki=True: {'Gakki': (1 if zenki else 2) if isinstance(zenki, bool) else zenki}
args_lnote = lambda lnote=True: {'vid': '05' if lnote else '03'}

args_general_list = {'module': 'General', 'tab': 2}
args_general_course = {'module': 'General', 'action': 'T0300'}

args_archive_list = {'module': 'Archive', 'action': 'ArchiveIndex', 'tab': 2, 'vid': 4}
args_archive_course = {'module': 'Archive', 'action': 'ArchiveIndex', 'SubAction': 'T0300'}
args_archive_course_old = {'module': 'Archive', 'action': 'KougiInfo', 'GakubuCD': 100, 'GakkaCD': 12}

# funcs
'''
インターネットと交互し、htmlを操り、データを導き出す部分である。
それぞれの関数の意味は名前の通りである。
'''
def get_department_list(year=2020, lang='JA'):
  url = 'http://www.ocw.titech.ac.jp/index.php?module=Archive&action=ArchiveIndex&GakubuCD=1&GakkaCD=311100&KeiCD=11&tab=2&focus=200&lang={}&Nendo={}&SubAction=T0200'.format(lang, year)
  html = driver.get_html_after_loaded(url)
  cats, cats_flat, excluded = extractor.get_department_list(html)
  print(excluded)
  return cats, cats_flat

def get_course_list(info, year=this_year, retry_limit=5):
  args_course = args_general_list if year >= this_year else args_archive_list
  if year >= this_year:
    if 'SubAction' in info:
      info = dict(info)
      info['action'] = info['SubAction']
      del info['SubAction']
  url = form_url(__url__=addr_default, **args_course, **args_year(year), **info)
  #print(url)
  
  tbls_ans = []
  for _ in range(retry_limit):
    html = driver.get_html_after_loaded(url)
    tbls_ans = extractor.get_course_list(html)
    if len(tbls_ans) > 0 and tbls_ans[0][3].startswith(str(year)):
      break
  
  return tbls_ans
  
def fetch_lecture_info_2(id_seed, year, en=False):
  '''
  Get the html return according to lecture id (and other information).
  lecture_info: id or (id, year) or (id, year, term)
  '''
  # request
  url0 = 'http://www.ocw.titech.ac.jp/index.php?module=General&action=T0300&JWC={year}{id}&lang={lang}'
  lang = 'JA' if not en else 'EN'
  url1 = url0.format(year=year, id=str(id_seed)[4:], lang=lang)
  url2 = url1 + '&vid=05'

  html = driver.get_html_after_loaded(url1)
  if html == '404':
    return None
  bf_base_page = BeautifulSoup(html, 'lxml')
  cnt_lname = bf_base_page.find_all(class_='page-title-area clearfix')[0]
  cnt_summ = bf_base_page.find_all(class_='gaiyo-data')[0]
  cnt_inner = bf_base_page.find_all(class_='right-inner')[0]
  
  html = driver.get_html_after_loaded(url2)
  bf_base_page = BeautifulSoup(html, 'lxml')
  try:
    cnt_note = bf_base_page.find_all(class_='right-inner')[0]
  except IndexError as e:
    cnt_note = bf_base_page.find_all(id='right-inner')[0]
  
  ans = (cnt_lname, cnt_summ, cnt_inner, cnt_note)
  return ans

# 開講元のリスト（つまり左側のあの赤いまたは青いもの）を取得する
def task1(outpath: str):
    d = get_department_list()
    e = get_department_list(lang='EN')
    pdump((d, e), outpath)

def task2(outpath: str):
  d = {
    '理学院': {'GakubuCD': 1},
    '工学院': {'GakubuCD': 2},
    '物質理工学院': {'GakubuCD': 3},
    '情報理工学院': {'GakubuCD': 4},
    '生命理工学院': {'GakubuCD': 5},
    '環境・社会理工学院': {'GakubuCD': 6},
    '教養科目群': {'GakubuCD': 7, 'GakkaCD': 370000},
    '初年次専門科目': {'GakubuCD': 10},
  }
  args_default = {'module': 'General', 'action': 'T0100'}
  args_la = {'module': 'General', 'action': 'T0200', 'tab': 2, 'focus': 100}
  
  # load cache
  try:
    cl1, cl2, nr1, nr2 = pload(outpath)
  except:
    cl1, cl2, nr1, nr2 = [], [], 0, 0
    pdump((cl1, cl2, nr1, nr2), outpath)
      
  for (i, target) in enumerate(d):
    if i < nr1:
      continue
    url = form_url(addr_default, 
                   **(d[target]), 
                   **(args_la if target == '教養科目群' else args_default), 
                   **args_lang())
    print(target, url)
    html = driver.get_html_after_loaded(url)
    lists = extractor.get_course_list(html)
    
    nr1 = i
    cl1 += lists
    
    backup_file(outpath)
    pdump((cl1, cl2, nr1, nr2), outpath)
  
  
def task3(inpath: str, outpath: str, start_year=2016, end_year=2022):
  clist, _, _, _ = pload(inpath)
  targets = [x[1][1][0] for x in clist]
  
  # initialize storage
  try:
    details, gshxd_code = pload(outpath)
  except:
    details = {} # these codes are successful in scraping
    gshxd_code = [] # these codes are yabai
    pdump((details, gshxd_code), outpath)
  
  def task3_sub1(detail: dict, dcode: Union[int, str], year: int):
    if year in detail:
      return
    detail_jp = extractor.parse_lecture_info(fetch_lecture_info_2(dcode, year))
    detail_en = extractor.parse_lecture_info(fetch_lecture_info_2(dcode, year, en=True))
    detail[year] = (detail_jp, detail_en)
  
  # error correction
  while gshxd_code:
    dcode = gshxd_code.pop()
    print(f'Retrying errored record {dcode} ({len(gshxd_code) + 1} remains)')
    
    detail = details[dcode] if dcode in details else {}
    for year in range(start_year, end_year + 1):
      try:
        task3_sub1(detail, dcode, year)
      except Exception as e:
        print(f'{dcode}, {year}')
        raise e
      
    details[dcode] = detail
    
    backup_file(outpath)
    pdump((details, gshxd_code), outpath)
    
  # normal process
  exit_tag = False
  for dcode in tqdm(targets):
    detail = details[dcode] if dcode in details else {}
    err_rec = False
    for year in range(start_year, end_year + 1):
      try:
        task3_sub1(detail, dcode, year)
      except KeyboardInterrupt:
        exit_tag = True
        break
      except Exception as e:
        if not err_rec:
          gshxd_code.append(dcode)
          err_rec = True
        print(f'ERROR {dcode} in year {year}: {e.message}')
        
    # to avoid incomplete records
    if exit_tag:
      print('Interrupted by user.')
      break
        
    details[dcode] = detail
    
    backup_file(outpath)
    pdump((details, gshxd_code), outpath)
    
    
  
if not driver.is_initialized():
  driver.init_driver()
  
if __name__ == '__main__':
  pass
  # a = fetch_lecture_info_2('202002964', 2017, en=True)
  #step1(start_year=2021, start_year_old=2021, omit_old_courses=True)
  #a = get_all_course_list(1)

File: Oocw.Crawler/Core/Crawler.cs
using System.Diagnostics;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

using AngleSharp;
using Oocw.Crawler.Utils;
using AngleSharp.Html.Dom;
using AngleSharp.Dom;
using Oocw.Crawler.Models;
using OpenQA.Selenium.DevTools;
using Oocw.Base;
using System.Web;
using Yaap;
using System.IO;

namespace Oocw.Crawler.Core;

using IDSS = IDictionary<string, string>;
using DSS = Dictionary<string, string>;
using IMDSS = ImmutableDictionary<string, string>;

using DictStrArgs = NestedDictionary<string, string?>;
using DictIntSyl = Dictionary<int, (SyllabusRecord?, SyllabusRecord?)>;

public class Crawler
{

    public const string ADDR_DEFAULT = "http://www.ocw.titech.ac.jp/index.php";

    public static readonly int THIS_YEAR = DateTime.Now.Year;

    public const int INNOVATION_YEAR = 2016;



    public readonly DriverWrapper Driver;

    public Crawler(DriverWrapper driver)
    {
        Driver = driver;
    }


    public static Func<string, DSS> args_lang = lang => new() { { "lang", lang } };

    public static Func<int, DSS> args_year = year => new() { { "Nendo", year.ToString() } };

    public static Func<bool, DSS> args_term = zenki => new() { { "Gakki", zenki ? "1" : "2" } };

    public static Func<bool, DSS> args_lnote = lnote => new() { { "vid", lnote ? "05" : "03" } };

    public static readonly IMDSS args_general_list = new DSS
    {
        ["module"] = "General",
        ["tab"] = "2"
    }.ToImmutableDictionary();

    public static readonly IMDSS args_general_course = new DSS
    {
        {"module", "General"},
        { "action","T0300"}
    }.ToImmutableDictionary();

    public static readonly IMDSS args_archive_list = new DSS
    {
        { "module", "Archive"},
        {  "action",  "ArchiveIndex"},
        {  "tab", "2"},
        { "vid",  "4"}
    }.ToImmutableDictionary();

    public static readonly IMDSS args_archive_course = new DSS
    {
        { "module", "Archive"},
        { "action","ArchiveIndex"},
        {"SubAction", "T0300"}
    }.ToImmutableDictionary();

    public static readonly IMDSS args_archive_course_old = new DSS
    {
        { "module", "Archive"},
        { "action", "KougiInfo"},
        { "GakubuCD", "100"},
        { "GakkaCD", "12"}
    }.ToImmutableDictionary();

    // funcs
    public (List<DictStrArgs>, DictStrArgs) GetDepartmentList(int year = 2020, string lang = "JA")
    {
        var url = $"http://www.ocw.titech.ac.jp/index.php?module=Archive&action=ArchiveIndex&GakubuCD=1&GakkaCD=311100&KeiCD=11&tab=2&focus=200&lang={lang}&Nendo={year}&SubAction=T0200";
        var html = Driver.GetHtmlAfterLoaded(url);
        var (cats, cats_flat, excluded) = DataExtractor.GetDepartmentList(html.Dom());
        Console.WriteLine(excluded);
        return (cats, cats_flat);
    }


    public IEnumerable<CourseRecord> GetCourseList(IDSS info, int year = 114514, int retry_limit = 5)
    {
        var args_course = year >= THIS_YEAR ? args_general_list : args_archive_list;

        if (year >= THIS_YEAR)
        {
            year = THIS_YEAR;
            if (info.ContainsKey("SubAction"))
            {
                var _info = new DSS(info)
                {
                    ["action"] = info["SubAction"]
                };
                _info.Remove("SubAction");
                info = _info;
            }
        }
        var url = ADDR_DEFAULT.AddUrlEncodedParameters(args_course, args_year(year), info);
        //print(url)
        var tbls_ans = new List<object>();
        foreach (var _ in Enumerable.Range(0, retry_limit))
        {
            var dom = Driver.GetHtmlAfterLoaded(url).Dom();
            var isCurrentYear = dom.QuerySelectorAll<IHtmlTableElement>("table.ranking-list").Where(
                x => x.Rows.Where(x => x.Cells[4].TextContent.NormalizeWebString().StartsWith(year.ToString())).Count() > 0)
                .Count() > 0;

            if (isCurrentYear)
            {
                return DataExtractor.GetCourseList(dom);
            }
        }
        return Enumerable.Empty<CourseRecord>();
    }

    // 
    //   Get the html return according to lecture id (and other information).
    //   lecture_info: id or (id, year) or (id, year, term)
    //   
    public (IElement?, IElement?, IElement?, IElement?) GetLectureInfo(int id_seed, int year, bool en = false)
    {
        // request
        var lang = !en ? "JA" : "EN";
        var url1 = $"http://www.ocw.titech.ac.jp/index.php?module=General&action=T0300&JWC={year}{id_seed.ToString()!.Substring(4)}&lang={lang}";
        var url2 = url1 + "&vid=05";
        var html = Driver.GetHtmlAfterLoaded(url1);
        if (html == "404")
        {
            return (null, null, null, null);
        }
        var dom = html.Dom();
        var nameTitleContainer = dom.QuerySelector(".page-title-area.clearfix");
        var summaryContainer = dom.QuerySelector(".gaiyo-data");
        var innerContainer = dom.QuerySelector(".right-inner");
        IElement? notesContainer = null;

        var hasSummary = dom.QuerySelector("li a.summary") != null;
        var hasNotes = dom.QuerySelector("li a.notes") != null;

        if (hasNotes) {
            html = Driver.GetHtmlAfterLoaded(url2);
            dom = html.Dom();
            notesContainer = dom.QuerySelector(".right-inner") ?? dom.QuerySelector("#right-inner");
        }
        
        var ans = (nameTitleContainer, summaryContainer, innerContainer, notesContainer);
        return ans;
    }

    /// <summary>
    /// 開講元のリスト（つまり左側のあの赤いまたは青いもの）を取得する
    /// </summary>
    /// <param name="outpath"></param>
    public void Task1(string outpath)
    {
        var d = GetDepartmentList();
        var e = GetDepartmentList(lang: "EN");
        FileUtils.Dump((d, e), outpath);
    }


    private static IDSS _k(int i) => new DSS() { ["GakubuCD"] = i.ToString() }.ToImmutableDictionary();
    public static readonly ImmutableList<(string, IDSS)> DepartmentCodes = new List<(string, IDSS)>()
    {
        ("理学院", _k(1)),
        ("工学院", _k(2)),
        ("物質理工学院", _k(3)),
        ("情報理工学院", _k(4)),
        ("生命理工学院", _k(5)),
        ("環境・社会理工学院", _k(6)),
        ("工系３学院", new DSS() { ["GakubuCD"]= "11", ["KamokuCD"]= "300001" }.ToImmutableDictionary()),
        ("教養科目群", new DSS() { ["GakubuCD"]= "7", ["GakkaCD"]= "370000" }.ToImmutableDictionary()),
        ("初年次専門科目", _k(10)),
    }.ToImmutableList();

    public static readonly IDSS args_default = new DSS
        {
            {"module","General"},
            {  "action","T0100"}
        }.ToImmutableDictionary();

    public static readonly IDSS args_la = new DSS
        {
            {"module", "General"},
            {"action", "T0200"},
            { "tab", "2"},
            {"focus", "100"}
        }.ToImmutableDictionary();

    public static readonly IDSS args_e3 = new DSS
        {
            {"module", "General"},
            {"action", "T0210"},
            { "tab", "2"},
            {"focus", "200"}
        }.ToImmutableDictionary();

    /// <summary>
    /// Get course list
    /// </summary>
    /// <param name="outPath"></param>
    public void Task2(string outPath)
    {
        List<CourseRecord> cl1, cl2;
        int nr1, nr2;

        // load cache
        try
        {
            (cl1, cl2, nr1, nr2) = FileUtils.Load<(List<CourseRecord>, List<CourseRecord>, int, int)>(outPath);
        }
        catch
        {
            cl1 = [];
            cl2 = [];
            nr1 = 0;
            nr2 = 0;
            FileUtils.Dump((cl1, cl2, nr1, nr2), outPath);
        }

        for (int i = 0; i < DepartmentCodes.Count; ++i)
        {
            var (target, dt) = DepartmentCodes[i];

            if (i < nr1)
                continue;
            var args = args_default;
            if (target == "教養科目群")
                args = args_la;
            else if (target == "工系３学院")
                args = args_e3;

            var url = ADDR_DEFAULT.AddUrlEncodedParameters(dt, args, args_lang("JA"));
            var urlEn = ADDR_DEFAULT.AddUrlEncodedParameters(dt, args, args_lang("EN"));
            Console.WriteLine($"{target} => {url}");

            var html = Driver.GetHtmlAfterLoaded(url);
            var lists = DataExtractor.GetCourseList(html.Dom());

            var htmlEn = Driver.GetHtmlAfterLoaded(urlEn);
            var listsEn = DataExtractor.GetCourseList(htmlEn.Dom());

            nr1 = i + 1;
            nr2 = i + 1;
            cl1.AddRange(lists);
            cl2.AddRange(listsEn);

            FileUtils.BackupFile(outPath);
            FileUtils.Dump((cl1, cl2, nr1, nr2), outPath);
        }
    }

    bool Task3Sub1(DictIntSyl detail, int courseCode, int year)
    {
        if (detail.ContainsKey(year))
        {
            return false;
        }
        var detail_jp = DataExtractor.ParseLectureInfo(GetLectureInfo(courseCode, year));
        var detail_en = DataExtractor.ParseLectureInfo(GetLectureInfo(courseCode, year, en: true));
        detail[year] = (detail_jp, detail_en);
        return true;
    }



    /// <summary>
    /// get course data
    /// </summary>
    /// <param name="inPath"></param>
    /// <param name="outPath"></param>
    /// <param name="startYear"></param>
    /// <param name="endYear"></param>
    /// <exception cref="Exception"></exception>
    public void Task3(string inPath, string outPath, int startYear = 2016, int endYear = 2022)
    {
        var (courseList, _, _, _) = FileUtils.Load<(List<CourseRecord>, List<CourseRecord>, int, int)>(inPath);
        var targets = courseList.Select(x =>
        {
            var idStr = HttpUtility.ParseQueryString(x.Url).Get("KougiCD", "JWC");
            var parsed = int.TryParse(idStr, out var idInt);
            if (!parsed)
            {
                Console.WriteLine($"Warning: Wrong format: {x.Url}");
                return -1;
            }
            return idInt;
        })
        .Where(x => x > 0)
        .ToList();
        // initialize storage

        Dictionary<int, DictIntSyl> details;
        List<int> erroredCodes;
        try
        {
            (details, erroredCodes) = FileUtils.Load<(Dictionary<int, DictIntSyl>, List<int>)>(outPath);
        }
        catch
        {
            details = [];
            erroredCodes = [];
            FileUtils.Dump((details, erroredCodes), outPath);
        }


        // error correction
        while (erroredCodes.Count > 0)
        {
            var errCode = erroredCodes.First();
            erroredCodes.RemoveAt(0);

            Console.WriteLine($"Retrying errored record {errCode} ({erroredCodes.Count + 1} remains)");

            if (!details.ContainsKey(errCode))
                details[errCode] = [];

            var detail = details[errCode];

            for (var year = startYear; year <= endYear; ++year)
            {
                Task3Sub1(detail, errCode, year);
            }

            FileUtils.BackupFile(outPath);
            FileUtils.Dump((details, erroredCodes), outPath);
        }

        // normal process
        var exitRequired = false;
        foreach (var courseCode in targets.Yaap())
        {
            if (!details.ContainsKey(courseCode))
                details[courseCode] = [];

            var detail = details[courseCode];
            var isCurrentCodeRecorded = false;
            var needDump = true;
            for (var year = startYear; year <= endYear; ++year)
            {
                try
                {
                    needDump = Task3Sub1(detail, courseCode, year);
                }
                catch (Exception e)
                {
                    if (!isCurrentCodeRecorded)
                    {
                        erroredCodes.Add(courseCode);
                        isCurrentCodeRecorded = true;
                    }
                    needDump = true;
                    Console.WriteLine($"ERROR {courseCode} in year {year}: {e}");
                }
            }
            // to avoid incomplete records
            if (exitRequired)
            {
                Console.WriteLine("Interrupted by user.");
                break;
            }
            details[courseCode] = detail;
            if (needDump) {
                FileUtils.BackupFile(outPath);
                FileUtils.Dump((details, erroredCodes), outPath);
            }
        }
    }


}
